core/data/src/main/kotlin/com/google/samples/apps/nowinandroid/core/data/util/ConnectivityManagerNetworkMonitor.kt

```diff
@@ -26,57 +26,71 @@ import android.net.NetworkRequest.Builder
 import android.os.Build.VERSION
 import android.os.Build.VERSION_CODES
 import androidx.core.content.getSystemService
+import androidx.tracing.Trace
+import androidx.tracing.trace
+import com.google.samples.apps.nowinandroid.core.network.Dispatcher
+import com.google.samples.apps.nowinandroid.core.network.NiaDispatchers.IO
 import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.channels.awaitClose
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.callbackFlow
 import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.flow.flowOn
 import javax.inject.Inject
 
 internal class ConnectivityManagerNetworkMonitor @Inject constructor(
     @ApplicationContext private val context: Context,
+    @Dispatcher(IO) private val ioDispatcher: CoroutineDispatcher,
 ) : NetworkMonitor {
     override val isOnline: Flow<Boolean> = callbackFlow {
-        val connectivityManager = context.getSystemService<ConnectivityManager>()
-        if (connectivityManager == null) {
-            channel.trySend(false)
-            channel.close()
-            return@callbackFlow
-        }
+        trace("NetworkMonitor.callbackFlow") {
+            val connectivityManager = context.getSystemService<ConnectivityManager>()
+            if (connectivityManager == null) {
+                channel.trySend(false)
+                channel.close()
+                return@callbackFlow
+            }
+
+            /**
+             * The callback's methods are invoked on changes to *any* network matching the [NetworkRequest],
+             * not just the active network. So we can simply track the presence (or absence) of such [Network].
+             */
+            val callback = object : NetworkCallback() {
 
-        /**
-         * The callback's methods are invoked on changes to *any* network matching the [NetworkRequest],
-         * not just the active network. So we can simply track the presence (or absence) of such [Network].
-         */
-        val callback = object : NetworkCallback() {
+                private val networks = mutableSetOf<Network>()
 
-            private val networks = mutableSetOf<Network>()
+                override fun onAvailable(network: Network) {
+                    networks += network
+                    channel.trySend(true)
+                }
 
-            override fun onAvailable(network: Network) {
-                networks += network
-                channel.trySend(true)
+                override fun onLost(network: Network) {
```

Comment:
Does onLost get called when the network remains, but it loses the NET_CAPABILITY_INTERNET capability?

Or do you need to listen to onCapabilitiesChanged() also?

Reply:
hm, I wasn't changing the implemntation here a tall, just wrapped it with the `trace` 😀